### PR TITLE
Fix to support update new key

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes are proposed in this pull request?

Fix to support update new key, for example, template key.

This is the dependency order. so we can review the bottom one first.

https://github.com/Alluxio/alluxio/pull/17305
↓
https://github.com/Alluxio/alluxio/pull/17780
↓
https://github.com/Alluxio/alluxio/pull/17850
↓
https://github.com/Alluxio/alluxio/pull/17856/

### Why are the changes needed?

Without this PR, we cannot update template key or any other key didn't exist in PropertyKeys.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

**Comment by lucyge2022** [comment]
@jiacheliu3  can you help take a look at this? I'm assuming propertykey doesn't allow unregistered templated key to be created for a reason.

File: tests/src/test/java/alluxio/client/cli/fsadmin/command/UpdateConfIntegrationTest.java [hunk 1]
```diff
@@ -32,12 +35,37 @@ public void unknownOption() {
 
   @Test
   public void updateUnknownKey() {
+    // Support update non-existing key
```
Comment (jiacheliu3) [inline tests/src/test/java/alluxio/client/cli/fsadmin/command/UpdateConfIntegrationTest.java:38]: typo?
```suggestion
    // Do not support update non-existing key
```

File: shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java [hunk 2]
```diff
@@ -49,14 +49,22 @@ public int run(CommandLine cl) throws IOException {
     int errorCode = 0;
     Map<PropertyKey, String> properties = new HashMap<>();
     for (String arg : cl.getArgList()) {
-      if (arg.contains("=")) {
+      if (!arg.contains("=")) {
+        System.err.printf("argument %s must contains \"=\"", arg);
```
Comment (jiacheliu3) [inline shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java:53]: return -1?

File: shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java [hunk 3]
```diff
@@ -49,14 +49,22 @@ public int run(CommandLine cl) throws IOException {
...
+          value = "";
+        } else if (kv.length > 2) {
+          // Normally, there must be only one "=", but there maybe more for some values
+          // of property contains "=", so just get the value from the part after first "=".
+          value = arg.substring(arg.indexOf("=") + 1);
+        } else {
+          value = kv[1];
         }
-        properties.put(PropertyKey.getOrBuildCustom(kv[0]), kv[1]);
+        properties.put(PropertyKey.Builder.stringBuilder(kv[0]).buildUnregistered(), value);
```
Comment (jiacheliu3) [inline shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java:67]: Do you intend to support setting all kinds of unknown keys in Alluxio? Or it's only a templated key? I think the logic here allows any key and it's probably not good?
```suggestion
        PropertyKey property = PropertyKey.Builder.stringBuilder(kv[0]).buildUnregistered();
        System.out.format("Setting property %s to %s%n", property, value);
        properties.put(PropertyKey.Builder.stringBuilder(kv[0]).buildUnregistered(), value);
```

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 4]
```diff
@@ -728,16 +728,23 @@ public Map<String, Boolean> updateConfiguration(Map<String, String> propertiesMa
...
+        // Build template key if possible, IllegalArgumentException for unknown key
+        key = PropertyKey.fromString(entry.getKey());
         if (Configuration.getBoolean(PropertyKey.CONF_DYNAMIC_UPDATE_ENABLED)
             && key.isDynamic()) {
-          Object oldValue = Configuration.get(key);
+          Object oldValue = null;
+          try {
+            oldValue = Configuration.get(key);
+          } catch (RuntimeException e) {
+            LOG.debug("Cannot get old value for the given key {}", key.getName());
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:740]: warning

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks for the work! PTAL

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java [hunk 5]
```diff
@@ -728,16 +728,23 @@ public Map<String, Boolean> updateConfiguration(Map<String, String> propertiesMa
     int successCount = 0;
     for (Map.Entry<String, String> entry : propertiesMap.entrySet()) {
       try {
-        PropertyKey key = PropertyKey.fromString(entry.getKey());
+        PropertyKey key;
+        // Build template key if possible, IllegalArgumentException for unknown key
+        key = PropertyKey.fromString(entry.getKey());
```
Comment (jiacheliu3) [inline core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:733]:
```suggestion
        // Build template key if possible, an IllegalArgumentException will be thrown for unknown key
        // The exception will be caught and client will receive the failure
        key = PropertyKey.fromString(entry.getKey());
```

File: shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java (hunk 3 above)
Comment (jiacheliu3) [reply shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java:67]: Ok got it, we send whatever key to the master, but the master only accepts templated keys. I've updated my suggested change above, PTAL.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
added two minor comments, mostly LGTM

File: core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java (hunk 4 above)
Comment (maobaolong) [reply core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java:741]: done

File: shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java (hunk 2 above)
Comment (maobaolong) [reply shell/src/main/java/alluxio/cli/fsadmin/command/UpdateConfCommand.java:53]: done

**Comment by maobaolong** [comment]
@jiacheliu3 Thanks for the previous review, I've addressed all of your suggestion, PTAL

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks for the work!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**